Miscellaneous small fixes and improvements for v5.0

swurzinger opened the pull request:
A collection of various smaller fixes and improvements for v5. Some of them have already been implemented recently on v5 and thus aren't included in the changes anymore:

 - show users, who are not in any group
 - avoid NPE in DateCast
 - product edit: don't hide images tab
 - more flexible product categories 
 - fix single delivery type loop, immediately hide selection dialog again
 - add to cart performance improvements
 - hide delivery types button when there's just a single one
 - fix payment dialog reactiveness
 - fix CORS/CSRF configuration
 - fix setup frontend navigation
 - fix price display, show full customer name
 - fix tabs selection via active property as used in pos-customers
 - POS: hide taxes button, when taxes are disabled
 - POS: fix product removal
 - POS: fix quantity change, minor typos
 - minor UI: avoid jumping scrollbar when spinner is active
 - fix layaway popup, fix pending order currency formatting
 - fix customer search, fix customer statement report
 - POS: fix customer selection popup
 - avoid endless loop when using invalid date
 - POS: fix cash register imports, add ts interface
 - replace deprecated vue lifecycle hooks
 - fix order products popup price
 - POS: fix cart-products switching, fix account full payment
 - add full customer name as calculated attribute, simplify name display
 - fix text readability (light theme)
 - fix layaway and manual payment
 - deduplicate payment status mapping
 - POS: fix initial queue waiting for never called resolve, cleanup cart buttons code
 - fix discount popup after extracting discout button code
 - fix typo in order settings dialog
 - fix product dialog close button
 - fix product price change input for default and advanced numpad
 - fix popup-closer for multiple layered popups
 - add missing popup-closer initialization for some popups
 - discount popup: allow decimal places
 - minor avatar code improvement
 - show full name in orders list
 - enable table cols caching
 - fix registers menu permissions
 - fix products list overflow on small screens

swurzinger commented:
Inteded to be sqashed before being merged. Didn't do it yet because maybe you want only a subset of the changes and some of them need to be reverted.

Blair2004 reviewed (changes requested):
There are personal changes that don't apply to the core project.

Blair2004 reviewed (changes requested):
This doesn't apply to the core project

Blair2004 reviewed (changes requested):
Doesn't apply to the core project

Blair2004 reviewed (changes requested):
What does this solves ?

Blair2004 reviewed (changes requested):
That's not a desired behavior. On grouped, products, we rather want to hide certain tabs.

Blair2004 reviewed (commented):
Not sure to understand this approach. Can you describe ?

Blair2004 commented on resources/ts/libraries/popup.ts:
```diff
@@ -160,18 +160,23 @@ export class Popup {
-        
-        setTimeout( () => {
+
+        if (immediately) {
```
This approach doesn't seems convenient.

swurzinger replied:
The problem I try to solve here is that the popup causes a delay of 250ms even if it isn't used (is immediately closed after being called, usually because there's nothing to choose from). The implementation changed slightly in a later commit due to your popup injection changes. Could you suggest a better approach I could implement?

Blair2004 commented on resources/ts/popups/ns-pos-order-type-popup.vue:
```diff
@@ -52,17 +53,17 @@ export default {
-        
-        this.typeSubscription   =   POS.types.subscribe( types => {
+
+        this.typeSubscription   =   POS.types.pipe(take(1)).subscribe( types => {
```
why would we take 1 types instead of all?

swurzinger replied:
The subscription here calls `select(type)` when there's only a single type and the select method then changes the types again with `types.next(...)`, leading to an endless loop. As the type subscription is used for initialization only I thought it's fine to only take the first and ignore anything else. It did work in my tests, but probably there's a better way to solve this. I'd be greatful about any ideas. :-)

Blair2004 reviewed (changes requested):
Not applicable to the core project

Blair2004 reviewed (changes requested):
This is not a great fit as it also applies to production.

Blair2004 reviewed (commented):
Building assets is part of the project as this avoids the need to install node, dependencies, and assets. A better solution is needed for taking this in consideration.

Blair2004 reviewed (changes requested):
we rather like the object implementation for cart button (instead of array).
Showing delivery button if 2 types are available doesn't consider the case if only Delivery is available as the only option.

swurzinger commented:
> Building assets is part of the project as this avoids the need to install node, dependencies, and assets. A better solution is needed for taking this in consideration.

I just removed all the build artifacts from the repo on my branch, reverted all those changes for this PR. So this PR shouldn't contain any changes to the assets.

> we rather like the object implementation for cart button (instead of array). Showing delivery button if 2 types are available doesn't consider the case if only Delivery is available as the only option.

You mean there is a case when there's a single delivery type but you still want to show the delivery type selection dialog? Do you maybe have an example? Because I don't understand yet why one would need that.

Blair2004 commented:
I believe those changes are no longer applicable to the current state of the app. A new PR might be necessary.
Thank you for the contribution.

swurzinger commented:
Most of the changes are still applicable to the latest NexoPos version, but this PR would need to be rebased. However, I chose to maintain a fork instead: https://github.com/swurzinger/NexoPOS/tree/stw/5-generic
Feel free to grab any fixes/improvements from there 😀
